lib/cloud-metadata/index.js

```diff
@@ -0,0 +1,197 @@
+const { getMetadataGcp } = require('./gcp')
+const { CallbackCoordination } = require('./callback-coordination')
+const DEFAULT_PROVIDER_CONFIG = {
+  aws: {
+    host: '169.254.169.254',
+    port: 80,
+    protocol: 'http'
+  },
+  azure: {
+    host: '169.254.169.254',
+    port: 80,
+    protocol: 'http'
+  },
+  gcp: {
+    host: 'metadata.google.internal',
+    port: 80,
+    protocol: 'http'
+  }
+}
+
+// we "ping" (in the colloquial sense, not the ICMP sense) the metadata
+// servers by having a listener that expects the underlying socket to
+// connect.  PING_TIMEOUT_MS control that time
+const PING_TIMEOUT_MS = 100
+
+// some of the metadata servers have a dedicated domain name.  This
+// value is added to the above socket ping to allow extra time for
+// the DNS to resolve
+const DNS_TIMEOUT_MS = 100
+
+// the metadata servers are HTTP services.  This value is
+// used as the timeout for the HTTP request that's
+// made.
+const HTTP_TIMEOUT_MS = 1000
+
+// timeout for the CallbackCoordination object -- this is a fallback to
+// account for a catastrophic error in the CallbackCoordination object
+const COORDINATION_TIMEOUT_MS = 3000
+
+class CloudMetadata {
+  constructor (agent) {
+    this.agent = agent
+  }
+
+  /**
+   * Fetches Cloud Metadata
+   *
+   * The module's main entry-point/primary method.  The getCloudMetadata function
+   * will query the cloud metadata servers and return (via a callback function)
+   * to final metadata object.  This function may be called with a single
+   * argument.
+   *
+   * getCloudMetadata(function(error, metadata){
+   *     //...
+   * })
+   *
+   * Or with two
+   *
+   * getCloudMetadata(config, function(error, metadata){
+   *     //...
+   * })
+   *
+   * The config parameter is an object that contains information on the
+   * metadata servers.  If omitted, the function will use the  module-global
+   * DEFAULT_PROVIDER_CONFIG values
+   */
+  getCloudMetadata (config, cb) {
+    // normalize arguments
+    if (!cb) {
+      cb = config
+      config = DEFAULT_PROVIDER_CONFIG
+    }
+    const fetcher = new CallbackCoordination(COORDINATION_TIMEOUT_MS)
+
+    if (this.shouldFetchGcp()) {
+      fetcher.schedule(function (fetcher) {
+        getMetadataGcp(
+          config.gcp.host,
+          config.gcp.port,
+          PING_TIMEOUT_MS + DNS_TIMEOUT_MS,
+          HTTP_TIMEOUT_MS,
+          config.gcp.protocol,
+          (error, result) => {
+            fetcher.recordResult(error, result)
+          }
+        )
+      })
+    }
+
+    if (this.shouldFetchAws()) {
+      // amazon has two metadata endpoints -- IMDSv1 and IMDSv2
+      // we schedule a callback to attempt to fetch from each
+      // one -- which callback succeeds first will win
+      fetcher.schedule(function (fetcher) {
+        getMetadataAwsV1(
+          config.aws.host,
+          config.aws.port,
+          PING_TIMEOUT_MS,
+          HTTP_TIMEOUT_MS,
+          config.aws.protocol,
```

Comment:
No particular reason, I usually don't reach for URL objects as my first choice and the multiple parameters with each part is just habit. I went a bit down the road of refactoring to see what it would take for a single URL paramater and I'm going to pass on this one, The way tests and the library are currently architected (Some of the metadata APIs require calls to multiple URLs, and tests require us to wait until the servers started to know what the port is) would make it awkward refactor and I'd rather avoid that when we're close to being done.